plugins/api/api_main.ml

```diff
@@ -1,39 +1,215 @@
+    else None
+
+  let is_api_at_path desc path = is_apiable_path @@ path / path_of_desc desc
+  let find_all_paths paths desc = List.filter ~f:(is_api_at_path desc) paths
+  let find_path paths desc = List.find ~f:(is_api_at_path desc) paths
+
+  let find_apis ?lang path =
+    let path' = match lang with
+      | Some lang -> with_lang path lang
+      | None -> path in
+    find_files path' |>
+    List.filter_map ~f:api_of_path
+
+end
+
+module Api = struct
+
+  type t = api
+  type op_err = [ `No_dst | `Fail of exn ]
+
+  let lang t = t.desc.lang
+  let name t = t.desc.name
+
+  let create lang path =
+    let path = Api_path.of_string path in
+    if Api_path.is_apiable_path path then
+      let name = Api_path.name_of_path path in
+      Some {desc = {lang; name}; path;}
+    else None
+
+  let cp dest api =
+    let dest' = Api_path.with_lang dest (lang api) in
+    ok_if_true (Api_path.exists dest) ~error:`No_dst >>= fun () ->
+    try_with
+      (fun () ->
+         let dest' = Api_path.to_string dest' in
+         FileUtil.mkdir dest';
+         FileUtil.cp [Api_path.to_string api.path] dest')
+
+  let rm path desc =
+    let path = Api_path.(path / path_of_desc desc) in
+    ok_if_true (Api_path.exists path) ~error:`No_dst >>= fun () ->
+    try_with (fun () -> FileUtil.rm [Api_path.to_string path])
+
+end
+
+module Api_error = struct
+
+  let exit_with () = exit 1
+
+  let api_exists path desc =
+    exit_with @@ eprintf
+      "Will not overwrite an API named %s at %s. Remove it before the updating.\n"
+      desc.name (Api_path.to_string path)
+
+  let api_not_exists ?path desc = match path with
+    | Some path ->
+      exit_with @@ eprintf "Can't find api %s:%s at %s\n"
+        desc.lang desc.name (Api_path.to_string path)
+    | None ->
+      exit_with @@ eprintf "Can't find api %s:%s\n"
+        desc.lang desc.name
+
+  let no_path api =
+    exit_with @@ eprintf
+      "There are not any writable paths for %s plugin\n" api.name
+
+  let fail descr exn =
+    exit_with @@ eprintf "%s: %s\n" descr (Exn.to_string exn)
+
+  let sanity_fail api api'  =
+    exit_with @@ eprintf
+      "Sanitization error: there are conflicting specifications for the same API(s):\n\
+       %s:%s at %s, %s\n\
+       Each API should be specified only once. Please resolve the conflict manually,\n\
+       either by removing the conflicting files or by renaming them.\n"
+      api.desc.lang api.desc.name
+      (Api_path.to_string api.path) (Api_path.to_string api'.path)
+end
+
+let add_api paths api =
+  match Api_path.find_path paths api.desc with
+  | Some path -> Api_error.api_exists path api.desc
+  | None ->
+    let rec add = function
+      | [] -> Api_error.no_path api.desc
+      | p :: paths' ->
+        match Api.cp p api with
+        | Ok () ->
+          info "wrote %s:%s to %s"
+            api.desc.lang api.desc.name (Api_path.to_string p)
+        | Error `No_dst ->
+          error "path not exists %s" (Api_path.to_string p);
```

Comment:
should be 
```
info "Api.cp: skipping a non-writable destination %a" Api.pp p"
```

It is not an error, or even a warning. There can be a plenty of paths that are writable, so there are no reason to panic.